Sources/Shared/AuthenticationController.swift

```diff
@@ -166,29 +177,38 @@ final public class AuthenticationController {
         return URI
     }
     
-    /**
-     Generate a URL to open the Vimeo code grant authorization page.  When opened in Safari, this page allows users to log into your application.
-     
-     - returns: the code grant authorization page URL
-     */
-    public func codeGrantAuthorizationURL() -> URL {
+    /// Generate a URL to open the Vimeo code grant authorization page.  When opened in Safari, this page allows users to log into your application.
+    ///
+    /// - Parameters:
+    ///   - name: Optional placeholder: when set, the user will see a placeholder name inside the name text field in Safari
+    ///   - email: Optional placeholder: when set, the user will see a placeholder email inside the email text field in Safari
+    ///   - action: Set this parameter to send user straight to join or login inside Safari, default is login
+    /// - Returns: the code grant authorization page URL
+    public func codeGrantAuthorizationURL(usingPlaceholdersForName name: String? = nil, andEmail email: String? = nil, showingInitialAuthorizationAction action: AuthorizationAction = .login) -> URL {
         let parameters = [Constants.ResponseTypeKey: Constants.CodeKey,
                           Constants.ClientIDKey: self.configuration.clientIdentifier,
                           Constants.RedirectURIKey: self.codeGrantRedirectURI,
                           Constants.ScopeKey: Scope.combine(self.configuration.scopes),
                           Constants.StateKey: type(of: self).state]
         
-        let urlString = self.configuration.baseUrl.appendingPathComponent(Constants.CodeGrantAuthorizationPath).absoluteString
+        let urlString = self.configuration.baseUrl.appendingPathComponent(Constants.CodeGrantAuthorizationPath)
+            .appendingQueries([
+                URLQueryItem(name: Constants.URLQueries.Email, value: email),
+                URLQueryItem(name: Constants.URLQueries.Name, value: name),
+                URLQueryItem(name: Constants.URLQueries.AuthorizationAction, value: action.rawValue)
+                ])
+            .absoluteString
+        
         
         var error: NSError?
         let urlRequest = VimeoRequestSerializer(appConfiguration: self.configuration).request(withMethod: VimeoClient.Method.GET.rawValue, urlString: urlString, parameters: parameters, error: &error)
         
-        guard let url = urlRequest.url, error == nil
-        else {
-            fatalError("Could not make code grant auth URL")
+        guard let codeGrantAuthorizationURL = urlRequest.url, error == nil
+            else {
+                fatalError("Could not make code grant auth URL")
```

Comment:
It's probably worth adding tests for invalid URL characters to see how this code path would behave in your unit tests. As far as I can see, if the call to `appendingQueries` failed somehow, the URL returned is the unmodified version of itself, which would also exclude `auth_action`. Is this acceptable?

Reply:
We should think about fallback, we would return unmodified URL and the user won't know why the placeholder is not shown. Any idea? I though on this solution because it keeps the same API. I can't get right now to a better solution

Reply:
So I though in case the user passed a name or email with invalid characters we should also make a fatalerror since I want to keep the same API and let the user know that some of the queries are invalid